lang/ruby/lib/avro/io.rb

```diff
@@ -31,6 +31,49 @@ def initialize(writers_schema, readers_schema)
       end
     end
 
+    def self.match_schemas(writers_schema, readers_schema)
+      w_type = writers_schema.type_sym
+      r_type = readers_schema.type_sym
+
+      # This conditional is begging for some OO love.
+      if w_type == :union || r_type == :union
+        return true
+      end
+
+      if w_type == r_type
+        return true if Schema::PRIMITIVE_TYPES_SYM.include?(r_type)
+
+        case r_type
+        when :record
+          return writers_schema.fullname == readers_schema.fullname
+        when :error
+          return writers_schema.fullname == readers_schema.fullname
+        when :request
+          return true
+        when :fixed
+          return writers_schema.fullname == readers_schema.fullname &&
+                 writers_schema.size == readers_schema.size
+        when :enum
+          return writers_schema.fullname == readers_schema.fullname
+        when :map
+          return writers_schema.values.type == readers_schema.values.type
+        when :array
+          return writers_schema.items.type == readers_schema.items.type
+        end
+      end
+
+      # Handle schema promotion
+      if w_type == :int && [:long, :float, :double].include?(r_type)
+        return true
+      elsif w_type == :long && [:float, :double].include?(r_type)
+        return true
+      elsif w_type == :float && r_type == :double
+        return true
+      end
+
+      return false
+    end
```

Comment:
this is unnecessary code churn.  In general do whatever you can to keep PRs smaller; I just spent 2 minutes reviewing this function before I realized you had just copy/pasted it out of a module.